Add info on attached/detached containers (#7376)
microsoft/FluidFramework: docs/content/docs/build/containers.md

### Container schema

You must define a schema that represents the structure of the data within your container. A schema can include `initialObjects` that are always available and types that can be dynamically created by the container at runtime. The same schema definition must be provided for creation and subsequent loading of the container. For more information on `initialObjects` and dynamic object creation see
[Data modeling]({{< relref "data-modeling.md" >}}).

This example schema defines two initial objects, and declares `SharedCell` and `SharedString` as shared object types that can be dynamically created at runtime.

```typescript {linenos=inline}
const containerSchema = {
    initialObjects: {
        layout: SharedMap,
        text: SharedString
    },
    dynamicObjectTypes: [ SharedCell, SharedString ],
};
```

### Creating a container

Containers are created from the service-specific client's `createContainer` function. You must provide a config that is specific to the service and a schema object that defines the container schema.

```typescript {linenos=inline,hl_lines=[7,8]}
const schema = {
    initialObjects: {
        layout: SharedMap,
    },
};

> const { container, containerServices } =

Reviewer: ```suggestion
const { container, services} =
```

Author: Aren't we calling these container services?

Reviewer: https://github.com/microsoft/FluidFramework/blob/600e75d4f845c383b2667d2d0ba320d176864e4b/packages/framework/tinylicious-client/src/TinyliciousClient.ts#L75

Reviewer: changed 2 weeks ago via https://github.com/microsoft/FluidFramework/pull/7254

Author: I see that, but if we are calling these things "container services" then I don't see the value in changing the variable name.

Reviewer: it's not a tuple return, `services` is the value returned via destructuring. `containerServices` will come back undefined/throw a type error

Author: Ahhhhh... I am once again thrown off by destructuring.

>     await client.createContainer(/*service config*/, schema);

Reviewer: create only takes a single schema param
```suggestion
    await client.createContainer(schema);
```

const containerId = await container.attach();
```

### Attaching a container

A newly created container is in a *detached* state. This is the point where you can create initial data to populate your

> DDSes if needed. A detached container is not connected to the Fluid service and no data is shared with other clients.

Author: ```suggestion
shared objects if needed. A detached container is not connected to the Fluid service and no data is shared with other clients.
```

In order to attach the container to a service, call its `attach` function. Once *attached*, the Fluid container is
connected to the Fluid service and can be loaded by other clients.

```typescript {linenos=inline,hl_lines=[10]}
const schema = {
    initialObjects: {
        layout: SharedMap,
    },
};

> const { container, containerServices } =

Reviewer: ```suggestion
const { container, services } =
```

    await client.createContainer(/*service config*/, schema);

const containerId = await container.attach();
```

### Loading a container

To load a container you must provide the service config as well as the exact same schema definition. The same container
schema is required on all subsequent loads or the container will not be loaded correctly.

Note that when loading an existing container, the container is already attached.

```typescript {linenos=inline}
const schema = {
    initialObjects: {
        layout: SharedMap,
    },
};

> const { container, containerServices } =

Reviewer: ```suggestion
const { container, services } =
```

    initialObjects: {
        layout: SharedMap,
        text: SharedString
    },
    dynamicObjectTypes: [ SharedCell, SharedString ],
};
```

### Creating a container

Containers are created from the service-specific client's `createContainer` function. You must provide a config that is specific to the service and a schema object that defines the container schema.

```typescript {linenos=inline,hl_lines=[7,8]}
const schema = {
    initialObjects: {
        layout: SharedMap,
    },
};

const { container, services } =
    await client.createContainer(schema);

const containerId = await container.attach();
```

### Attaching a container

A newly created container is in a *detached* state. This is the point where you can create initial data to populate your
shared objects if needed. A detached container is not connected to the Fluid service and no data is shared with other clients.

In order to attach the container to a service, call its `attach` function. Once *attached*, the Fluid container is
connected to the Fluid service and can be loaded by other clients.

Note that once attached, a container cannot be detached. Attach is a one-way operation. When loading an existing

> container, the loaded container is always attached.

Reviewer: bit duplicative of line 83

```typescript {linenos=inline,hl_lines=[10]}
const schema = {
    initialObjects: {
        layout: SharedMap,
    },
};

const { container, services } =
    await client.createContainer(schema);

const containerId = await container.attach();
```

### Loading a container

> To load a container you must provide the service config as well as the exact same schema definition. The same container

Reviewer: getContainer always takes an 'id' as the first param right? Not a service config? That config goes into the client creation.

Author: ```suggestion
To load a container you must provide the `id` of the container to load as well as the exact same schema definition used when creating the container. The same container
```
